File: Kotlin/ByRound/2159/2159_D1_Inverse_Minimum_Partition_Easy_Version.kt
// Problem: CF 2159 D1 - Inverse Minimum Partition (Easy Version)
// https://codeforces.com/contest/2159/problem/D1

/*
D1. Inverse Minimum Partition (Easy Version)

Algorithm:
This solution uses dynamic programming with a TreeMap to efficiently find the minimum cost to partition the array.
- For each position k, we compute dp[k] as the minimum cost to partition the first k elements.
- The key idea is to maintain a TreeMap of values and their indices where we can make a cut.
- For each element, we remove entries from the TreeMap that are >= current element.
- Then we query for valid previous positions by looking up the largest value < ceilDiv(current_element, cost).
- Time Complexity: O(n * log n) per test case due to TreeMap operations and outer loop.
- Space Complexity: O(n) for dp array and TreeMap.

Techniques:
- Dynamic Programming
- TreeMap for efficient range queries and updates
- Ceiling division calculation

*/
import java.util.TreeMap
import kotlin.math.min

fun main() {
    repeat(readln().toInt()) {
        val n = readln().toInt()
        val xs = listOf(0L) + readln().split(" ").map(String::toLong)
        val treeMap = TreeMap<Long, Int>() // Stores (value, index) pairs, sorted by value
        treeMap[0L] = 0
        val dp = IntArray(n + 1) { n + 1 } // dp[i] represents min cost to partition first i elements
        dp[0] = 0
        for (k in 1..n) {
            while (treeMap.lastKey() >= xs[k]) { // Remove values greater than or equal to current element
                treeMap.remove(treeMap.lastKey())
            }
            treeMap[xs[k]] = k // Insert current element and its index into TreeMap
            for (cost in 1..5) {
                val prev = treeMap.lowerEntry(ceilDiv(xs[k], cost.toLong())).value // Find largest entry < ceilDiv(xs[k], cost)
                dp[k] = min(dp[k], dp[prev] + cost) // Update dp with minimum cost
            }
        }
        println(dp[n])
    }
}

fun ceilDiv(x: Long, y: Long) = (x + y - 1L) / y


// https://github.com/VaHiX/CodeForces/

File: CPP/ByRound/2159/2159_D1_Inverse_Minimum_Partition_Easy_Version.cpp
// Problem: CF 2159 D1 - Inverse Minimum Partition (Easy Version)
// https://codeforces.com/contest/2159/problem/D1

/*
Problem: Inverse Minimum Partition (Easy Version)

This problem involves partitioning a sequence to minimize the total cost,
where the cost of a subsequence is defined as ceil(k / min_element) where k is the length.

Algorithm:
- Preprocess rightmost minimum indices using dynamic programming.
- Use another DP to compute minimal cost partition.
- For each position, try multiplying by 2 and 3 to find valid next partitions.

Time Complexity: O(n * log(max(a[i]))) per test case
Space Complexity: O(n)

Techniques Used:
- Dynamic Programming
- Preprocessing with suffix minimums
- Binary search using upper_bound
*/

#include <algorithm>
#include <initializer_list>
#include <iostream>
#include <vector>

using namespace std;
int main() {
  ios_base::sync_with_stdio(0);
  cin.tie(0);
  int T;
  cin >> T;
  while (T--) {
    int n;
    cin >> n;
    vector<long long> a(n);
    for (auto &i : a)
      cin >> i;
    vector<int> rightmost_min(n + 1);
    rightmost_min[n] = n; // Base case: no elements to the right
    rightmost_min[n - 1] = n - 1; // Last element's rightmost min is itself
    for (int i = n - 2; i >= 0; i--) {
      if (a[i] < a[rightmost_min[i + 1]]) {
        rightmost_min[i] = i; // Current element is smaller than rightmost minimum, so it is new minimum
      } else {
        rightmost_min[i] = rightmost_min[i + 1]; // Otherwise inherit the rightmost minimum index
      }
    }
    vector<long long> b;
    int idx = rightmost_min[0];
    while (idx < n) {
      b.push_back(a[idx]); // Build the sequence of rightmost minimums
      idx = rightmost_min[idx + 1]; // Move to next partition point
    }
    int m = b.size();
    vector<int> dp(m + 1);
    dp[m] = 0; // Base case: no elements left, cost is zero
    dp[m - 1] = 1; // If one element remains, cost is 1
    for (int i = m - 2; i >= 0; i--) {
      dp[i] = 1 + dp[i + 1]; // Take current as a separate part, and continue from next
      for (auto &mult : {2, 3}) {
        int j = upper_bound(b.begin(), b.end(), mult * b[i]) - b.begin(); // Find first value > mult * b[i]
        dp[i] = min(dp[i], mult + dp[j]); // Try to extend current range by multiplying with 2 or 3
      }
    }
    cout << dp[0] << "\n"; // Final result for this test case
  }
}


// https://github.com/VaHiX/CodeForces/